File: platform/polyglot/langgraph-flow-serializer/mjs/src/adapters/langflow.adapter.ts
/**
 * Langflow dialect adapter.
 *
 * Langflow uses `data.node.template` with per-field objects of shape
 * `{ type, value, show, ... }`. The LangChain class name lives in `data.type`.
 *
 * Detection heuristic: nodes have `data.node.template` present as an object.
 */

import type { AIWorkflow, AINode, AIEdge } from "../types.js";
import type { DialectAdapter } from "./adapter.js";

interface LangflowTemplateField {
  type?: string;
  value?: unknown;
  show?: boolean;
  [key: string]: unknown;
}

interface LangflowNodeInner {
  template?: Record<string, LangflowTemplateField>;
  display_name?: string;
  description?: string;
  [key: string]: unknown;
}

interface LangflowNodeData {
  type?: string;
  node?: LangflowNodeInner;
  category?: string;
  [key: string]: unknown;
}

interface LangflowNode {
  id: string;
  type?: string;
  position?: { x: number; y: number };
  data?: LangflowNodeData;
  [key: string]: unknown;
}

interface LangflowEdge {
  id: string;
  source: string;
  sourceHandle?: string;
  target: string;
  targetHandle?: string;
  type?: string;
  [key: string]: unknown;
}

interface LangflowFlow {
  id?: string;
  name?: string;
  description?: string;
  viewport?: { x: number; y: number; zoom: number };
  nodes?: LangflowNode[];
  edges?: LangflowEdge[];
  [key: string]: unknown;
}

export const langflowAdapter: DialectAdapter = {
  detect(json: unknown): boolean {
    if (typeof json !== "object" || json === null) return false;
    const flow = json as LangflowFlow;
    const nodes = flow.nodes;
    if (!Array.isArray(nodes) || nodes.length === 0) return false;

    const firstNode = nodes[0];
    if (typeof firstNode?.data !== "object" || firstNode.data === null) return false;
    const data = firstNode.data as LangflowNodeData;
    return (
      typeof data.node === "object" &&
      data.node !== null &&
      typeof (data.node as LangflowNodeInner).template === "object"
    );
  },

  fromDialect(dialectJson: unknown): AIWorkflow {
    const flow = dialectJson as LangflowFlow;

    const nodes: AINode[] = (flow.nodes ?? []).map((ln): AINode => {
      const data = ln.data ?? {};
      const inner = data.node ?? {};
      const template = inner.template ?? {};

      // Flatten template: extract value from each field
      const inputs: Record<string, unknown> = {};
      for (const [key, field] of Object.entries(template)) {
        inputs[key] = field.value ?? null;
      }

      return {
        id: ln.id,
        type: ln.type ?? "customNode",
        position: ln.position ?? { x: 0, y: 0 },
        data: {
          nodeType: data.type ?? "",
          category: data.category ?? "Uncategorized",
          name: inner.display_name ?? data.type,
          inputs,
        },
      };
    });

    const edges: AIEdge[] = (flow.edges ?? []).map((le): AIEdge => ({
      id: le.id,
      source: le.source,
      sourceHandle: le.sourceHandle ?? `${le.source}-output`,
      target: le.target,
      targetHandle: le.targetHandle ?? `${le.target}-input`,
      ...(le.type ? { type: le.type } : {}),
    }));

    return {
      ...(flow.id ? { id: flow.id } : {}),
      name: flow.name ?? "Untitled Langflow",
      ...(flow.description ? { description: flow.description } : {}),
      ...(flow.viewport ? { viewport: flow.viewport } : {}),
      nodes,
      edges,
    };
  },

  toDialect(workflow: AIWorkflow): LangflowFlow {
    const nodes: LangflowNode[] = workflow.nodes.map((n): LangflowNode => {
      // Reconstruct the template structure with minimal metadata
      const template: Record<string, LangflowTemplateField> = {};
      for (const [key, value] of Object.entries(n.data.inputs)) {
        template[key] = {
          type: typeof value === "number" ? "float" : "str",
          value,
          show: true,
        };
      }

      return {
        id: n.id,
        type: n.type,
        position: n.position,
        data: {
          type: n.data.nodeType,
          category: n.data.category,
          node: {
            display_name: n.data.name ?? n.data.nodeType,
            template,
          },
        },
      };
    });

    const edges: LangflowEdge[] = workflow.edges.map((e): LangflowEdge => ({
      id: e.id,
      source: e.source,
      sourceHandle: e.sourceHandle,
      target: e.target,
      targetHandle: e.targetHandle,
      ...(e.type ? { type: e.type } : {}),
    }));

    return {
      ...(workflow.id ? { id: workflow.id } : {}),
      name: workflow.name,
      ...(workflow.description ? { description: workflow.description } : {}),
      ...(workflow.viewport ? { viewport: workflow.viewport } : {}),
      nodes,
      edges,
    };
  },
};

File: platform/polyglot/langgraph-flow-serializer/py/langgraph_flow_serializer/adapters/langflow_adapter.py
"""
Langflow dialect adapter.

Mirrors mjs/src/adapters/langflow.adapter.ts.

Langflow uses data.node.template with per-field objects:
  { "type": "str", "value": "...", "show": true }

Detection heuristic: first node has data.node.template as a dict.
"""

from __future__ import annotations

from typing import Any

from langgraph_flow_serializer.adapters.base import DialectAdapter
from langgraph_flow_serializer.types import AIEdge, AINode, AINodeData, AIViewport, AIWorkflow


class LangflowAdapter(DialectAdapter):
    def detect(self, json: Any) -> bool:
        if not isinstance(json, dict):
            return False
        nodes = json.get("nodes")
        if not isinstance(nodes, list) or len(nodes) == 0:
            return False
        first = nodes[0]
        if not isinstance(first, dict):
            return False
        data = first.get("data")
        if not isinstance(data, dict):
            return False
        inner = data.get("node")
        return isinstance(inner, dict) and isinstance(inner.get("template"), dict)

    def from_dialect(self, dialect_json: Any) -> AIWorkflow:
        nodes: list[AINode] = []
        for ln in dialect_json.get("nodes", []):
            data = ln.get("data") or {}
            inner = data.get("node") or {}
            template = inner.get("template") or {}

            # Flatten: extract value from each template field
            inputs: dict = {}
            for key, field in template.items():
                inputs[key] = field.get("value") if isinstance(field, dict) else field

            node_data = AINodeData(
                node_type=data.get("type", ""),
                category=data.get("category", "Uncategorized"),
                name=inner.get("display_name") or data.get("type"),
                inputs=inputs,
            )
            nodes.append(
                AINode(
                    id=ln["id"],
                    type=ln.get("type", "customNode"),
                    position=ln.get("position") or {"x": 0, "y": 0},
                    data=node_data,
                )
            )

        edges: list[AIEdge] = []
        for le in dialect_json.get("edges", []):
            edges.append(
                AIEdge(
                    id=le["id"],
                    source=le["source"],
                    source_handle=le.get("sourceHandle") or f"{le['source']}-output",
                    target=le["target"],
                    target_handle=le.get("targetHandle") or f"{le['target']}-input",
                    type=le.get("type") or None,
                )
            )

        vp_raw = dialect_json.get("viewport")
        viewport = (
            AIViewport(x=vp_raw["x"], y=vp_raw["y"], zoom=vp_raw["zoom"])
            if isinstance(vp_raw, dict)
            else None
        )

        return AIWorkflow(
            id=dialect_json.get("id") or None,
            name=dialect_json.get("name") or "Untitled Langflow",
            description=dialect_json.get("description") or None,
            viewport=viewport,
            nodes=nodes,
            edges=edges,
        )

    def to_dialect(self, workflow: AIWorkflow) -> dict:
        nodes = []
        for n in workflow.nodes:
            template = {}
            for key, value in n.data.inputs.items():
                if isinstance(value, (int, float)):
                    field_type = "float"
                else:
                    field_type = "str"
                template[key] = {"type": field_type, "value": value, "show": True}

            nodes.append(
                {
                    "id": n.id,
                    "type": n.type,
                    "position": n.position,
                    "data": {
                        "type": n.data.node_type,
                        "category": n.data.category,
                        "node": {
                            "display_name": n.data.name or n.data.node_type,
                            "template": template,
                        },
                    },
                }
            )

        edges = [
            {
                "id": e.id,
                "source": e.source,
                "sourceHandle": e.source_handle,
                "target": e.target,
                "targetHandle": e.target_handle,
                **({"type": e.type} if e.type else {}),
            }
            for e in workflow.edges
        ]

        result: dict = {"name": workflow.name, "nodes": nodes, "edges": edges}
        if workflow.id:
            result["id"] = workflow.id
        if workflow.description:
            result["description"] = workflow.description
        if workflow.viewport:
            result["viewport"] = {
                "x": workflow.viewport.x,
                "y": workflow.viewport.y,
                "zoom": workflow.viewport.zoom,
            }
        return result
